File: forecast_engine/generate-demo-data.js
/**
 * Demo Data Generator for DemandIQ
 * Generates CSV files with 120+ SKUs, 36 months of sales history, and promotions
 * Usage: node generate-demo-data.js
 */

const fs = require('fs');
const path = require('path');

function generateProducts(numSkus = 120) {
  const categories = ['Electronics', 'Apparel', 'Food & Beverage', 'Home & Garden', 'Sports', 'Automotive'];
  const brands = ['TechCorp', 'StyleMax', 'NutriHub', 'HomeFlow', 'AthleteZone', 'AutoPro'];
  const subcats = ['Premium', 'Standard', 'Budget', 'Professional', 'Casual'];

  const products = [];
  let skuCounter = 1000;

  const baseProducts = [
    { sku: 'EL-001', name: 'Wireless Headphones Pro', cat: 'Electronics', subcat: 'Audio', brand: 'SoundCore', price: 149.99, lt: 14, abc: 'A', xyz: 'X' },
    { sku: 'EL-002', name: 'Smart Watch Series 5', cat: 'Electronics', subcat: 'Wearables', brand: 'TechFit', price: 299.99, lt: 21, abc: 'A', xyz: 'Y' },
    { sku: 'EL-003', name: 'Bluetooth Speaker Mini', cat: 'Electronics', subcat: 'Audio', brand: 'SoundCore', price: 79.99, lt: 10, abc: 'B', xyz: 'X' },
    { sku: 'AP-001', name: 'Performance Running Shoes', cat: 'Apparel', subcat: 'Footwear', brand: 'StepFast', price: 119.99, lt: 30, abc: 'A', xyz: 'Y' },
    { sku: 'FB-001', name: 'Premium Coffee Blend', cat: 'Food & Beverage', subcat: 'Coffee', brand: 'BeanMaster', price: 24.99, lt: 7, abc: 'A', xyz: 'X' },
  ];

  for (const p of baseProducts) {
    products.push({
      sku: p.sku,
      product_name: p.name,
      category: p.cat,
      subcategory: p.subcat,
      brand: p.brand,
      unit_price: p.price,
      lead_time_days: p.lt,
      abc_class: p.abc,
      xyz_class: p.xyz,
    });
  }

  // Generate additional SKUs
  for (let catIdx = 0; catIdx < categories.length; catIdx++) {
    for (let i = 0; i < 15; i++) {
      const catPrefix = categories[catIdx].substring(0, 2).toUpperCase();
      const sku = `${catPrefix}-${String(skuCounter++).padStart(3, '0')}`;

      const basePrice = 30 + Math.random() * 250;
      const leadTime = 5 + Math.floor(Math.random() * 30);
      const abcHash = (catIdx * 17 + i * 13) % 100;
      const abc_class = abcHash < 80 ? 'A' : abcHash < 95 ? 'B' : 'C';
      const xyz_class = ['X', 'Y', 'Z'][Math.floor(Math.random() * 3)];

      products.push({
        sku,
        product_name: `${categories[catIdx]} Product ${i + 1}`,
        category: categories[catIdx],
        subcategory: subcats[Math.floor(Math.random() * subcats.length)],
        brand: brands[catIdx],
        unit_price: Math.round(basePrice * 100) / 100,
        lead_time_days: leadTime,
        abc_class,
        xyz_class,
      });
    }
  }

  return products;
}

function generateSalesHistory(products, months = 36) {
  const records = [];
  const now = new Date();
  const promotionMonths = [9, 10, 11, 4];

  for (const product of products) {
    const { sku, unit_price: price } = product;
    const skuHash = sku.split('').reduce((a, c) => a + c.charCodeAt(0), 0);
    const baseUnits = 200 + (skuHash % 800);

    for (let monthOffset = months; monthOffset >= 0; monthOffset--) {
      const date = new Date(now.getFullYear(), now.getMonth() - monthOffset, 1);
      const monthIdx = date.getMonth();

      const holidayPeak = monthIdx >= 10 || monthIdx <= 0 ? 1.4 : 1.0;
      const summerDip = monthIdx >= 6 && monthIdx <= 7 ? 0.7 : 1.0;
      const springUpswing = monthIdx >= 2 && monthIdx <= 4 ? 1.15 : 1.0;
      const baseSeasonal = holidayPeak * summerDip * springUpswing;

      const seasonal = Math.sin(((monthIdx - 2) / 12) * 2 * Math.PI) * (baseUnits * 0.3) * baseSeasonal;
      const trend = (36 - monthOffset) * (baseUnits * 0.015);
      const noise = (Math.random() - 0.5) * (baseUnits * 0.12);

      const promoChance = promotionMonths.includes(monthIdx) ? 0.25 : 0.08;
      const promoFlag = Math.random() < promoChance;

      const stockoutChance = promoFlag ? 0.08 : 0.03;
      const stockoutFlag = Math.random() < stockoutChance;

      let units = Math.max(10, Math.round(baseUnits + seasonal + trend + noise));

      if (promoFlag) {
        units = Math.round(units * (1.25 + Math.random() * 0.3));
      }

      if (stockoutFlag) {
        units = Math.round(units * 0.4);
      }

      records.push({
        sku,
        date: date.toISOString().split('T')[0],
        units_sold: units,
        revenue: Math.round(units * price * 100) / 100,
        promotion_flag: promoFlag ? 1 : 0,
        stockout_flag: stockoutFlag ? 1 : 0,
      });
    }
  }

  return records;
}

function generatePromotions(products) {
  const promotions = [];
  const now = new Date();

  for (const product of products) {
    const { sku, abc_class } = product;

    // A-class products get more promotions
    if (abc_class === 'A' && Math.random() < 0.6) {
      const eventStart = new Date(now.getFullYear(), 9, 15); // October 15
      const eventEnd = new Date(now.getFullYear(), 11, 31); // Dec 31
      promotions.push({
        sku,
        event_name: `Q4 Holiday Campaign - ${sku}`,
        event_type: 'holiday',
        start_date: eventStart.toISOString().split('T')[0],
        end_date: eventEnd.toISOString().split('T')[0],
        discount_percent: Math.round((15 + Math.random() * 25) * 100) / 100,
        expected_lift_percent: Math.round((25 + Math.random() * 40) * 100) / 100,
        status: 'planned',
      });
    }

    // Black Friday
    if (Math.random() < 0.7) {
      const bfDate = new Date(now.getFullYear(), 10, 25); // November 25
      promotions.push({
        sku,
        event_name: `Black Friday Sale - ${sku}`,
        event_type: 'flash_sale',
        start_date: new Date(bfDate.getTime() - 2 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        end_date: new Date(bfDate.getTime() + 2 * 24 * 60 * 60 * 1000).toISOString().split('T')[0],
        discount_percent: Math.round((30 + Math.random() * 20) * 100) / 100,
        expected_lift_percent: Math.round((50 + Math.random() * 50) * 100) / 100,
        status: 'planned',
      });
    }
  }

  return promotions;
}

function toCsv(data) {
  if (!data || data.length === 0) return '';
  const headers = Object.keys(data[0]);
  const rows = data.map(row =>
    headers.map(h => {
      const val = row[h];
      if (typeof val === 'string' && (val.includes(',') || val.includes('"'))) {
        return `"${val.replace(/"/g, '""')}"`;
      }
      return val;
    }).join(',')
  );
  return [headers.join(','), ...rows].join('\n');
}

// Main execution
console.log('Generating extended product catalog...');
const products = generateProducts(120);
const productsCSV = toCsv(products);
fs.writeFileSync(path.join(__dirname, 'demo_products.csv'), productsCSV);
console.log(`✓ Generated ${products.length} products -> demo_products.csv`);

console.log('Generating 36 months of sales history...');
const sales = generateSalesHistory(products, 36);
const salesCSV = toCsv(sales);
fs.writeFileSync(path.join(__dirname, 'demo_sales_history.csv'), salesCSV);
console.log(`✓ Generated ${sales.length} sales records -> demo_sales_history.csv`);

console.log('Generating promotion events...');
const promos = generatePromotions(products);
const promosCSV = toCsv(promos);
fs.writeFileSync(path.join(__dirname, 'demo_promotions.csv'), promosCSV);
console.log(`✓ Generated ${promos.length} promotion events -> demo_promotions.csv`);

console.log(`
Dataset Summary:
  Products: ${products.length}
  Sales Records: ${sales.length} (${Math.floor(sales.length / products.length)} months average)
  Promotion Events: ${promos.length}

Files generated in forecast_engine/:
  - demo_products.csv
  - demo_sales_history.csv
  - demo_promotions.csv
`);

File: forecast_engine/generate_demo_data.py
"""
Generate comprehensive demo dataset for demand planning platform
Output: CSV files for import into the application
"""

import pandas as pd
import numpy as np
from datetime import datetime, timedelta
import random

def generate_products(num_skus=120):
    """Generate extended product catalog with 120+ SKUs"""
    categories = ['Electronics', 'Apparel', 'Food & Beverage', 'Home & Garden', 'Sports', 'Automotive']
    brands = ['TechCorp', 'StyleMax', 'NutriHub', 'HomeFlow', 'AthleteZone', 'AutoPro']
    subcats = ['Premium', 'Standard', 'Budget', 'Professional', 'Casual']

    products = []
    sku_counter = 1000

    base_products = [
        ('EL-001', 'Wireless Headphones Pro', 'Electronics', 'Audio', 'SoundCore', 149.99, 14, 'A', 'X'),
        ('EL-002', 'Smart Watch Series 5', 'Electronics', 'Wearables', 'TechFit', 299.99, 21, 'A', 'Y'),
        ('EL-003', 'Bluetooth Speaker Mini', 'Electronics', 'Audio', 'SoundCore', 79.99, 10, 'B', 'X'),
        ('AP-001', 'Performance Running Shoes', 'Apparel', 'Footwear', 'StepFast', 119.99, 30, 'A', 'Y'),
        ('FB-001', 'Premium Coffee Blend', 'Food & Beverage', 'Coffee', 'BeanMaster', 24.99, 7, 'A', 'X'),
    ]

    products.extend([{
        'sku': sku, 'product_name': name, 'category': cat, 'subcategory': subcat,
        'brand': brand, 'unit_price': price, 'lead_time_days': lt, 'abc_class': abc, 'xyz_class': xyz
    } for sku, name, cat, subcat, brand, price, lt, abc, xyz in base_products])

    # Generate additional SKUs
    for cat_idx, category in enumerate(categories):
        for i in range(15):
            cat_prefix = category[:2].upper()
            sku = f"{cat_prefix}-{sku_counter:03d}"
            sku_counter += 1

            base_price = 30 + random.random() * 250
            lead_time = 5 + random.randint(0, 30)
            abc_hash = (cat_idx * 17 + i * 13) % 100
            abc_class = 'A' if abc_hash < 80 else ('B' if abc_hash < 95 else 'C')
            xyz_class = random.choice(['X', 'Y', 'Z'])

            products.append({
                'sku': sku,
                'product_name': f"{category} Product {i+1}",
                'category': category,
                'subcategory': random.choice(subcats),
                'brand': brands[cat_idx],
                'unit_price': round(base_price, 2),
                'lead_time_days': lead_time,
                'abc_class': abc_class,
                'xyz_class': xyz_class
            })

    return pd.DataFrame(products)

def generate_sales_history(products_df, months=36):
    """Generate 36 months of realistic sales history with seasonality, promotions, stockouts"""
    sales_records = []
    now = datetime.now()

    # Promotion calendar
    promotion_months = [9, 10, 11, 4]  # Oct, Nov, Dec, May (0-indexed Sep, Oct, Nov, Apr)
    black_friday_weeks = [47, 48]

    for _, product in products_df.iterrows():
        sku = product['sku']
        price = product['unit_price']

        # SKU-specific base demand derived from hash
        sku_hash = sum(ord(c) for c in sku)
        base_units = 200 + (sku_hash % 800)

        for month_offset in range(months, -1, -1):
            date = now - timedelta(days=30*month_offset)
            month_idx = date.month - 1

            # Multi-harmonic seasonality
            holiday_peak = 1.4 if (month_idx >= 10 or month_idx <= 0) else 1.0
            summer_dip = 0.7 if 6 <= month_idx <= 7 else 1.0
            spring_upswing = 1.15 if 2 <= month_idx <= 4 else 1.0
            base_seasonal = holiday_peak * summer_dip * spring_upswing

            seasonal = np.sin(((month_idx - 2) / 12) * 2 * np.pi) * (base_units * 0.3) * base_seasonal
            trend = (36 - month_offset) * (base_units * 0.015)
            noise = (random.random() - 0.5) * (base_units * 0.12)

            # Promotion probability
            promo_chance = 0.25 if month_idx in promotion_months else 0.08
            promo_flag = random.random() < promo_chance

            # Stockout probability
            stockout_chance = 0.08 if promo_flag else 0.03
            stockout_flag = random.random() < stockout_chance

            units = max(10, int(base_units + seasonal + trend + noise))

            if promo_flag:
                units = int(units * (1.25 + random.random() * 0.3))

            if stockout_flag:
                units = int(units * 0.4)

            sales_records.append({
                'sku': sku,
                'date': date.strftime('%Y-%m-%d'),
                'units_sold': units,
                'revenue': units * price,
                'promotion_flag': int(promo_flag),
                'stockout_flag': int(stockout_flag)
            })

    return pd.DataFrame(sales_records)

def generate_promotions(products_df):
    """Generate promotion event calendar"""
    promotions = []
    now = datetime.now()

    event_types = ['seasonal', 'clearance', 'holiday', 'flash_sale', 'bundle']

    for _, product in products_df.iterrows():
        sku = product['sku']

        # A-class products get more promotions
        if product['abc_class'] == 'A' and random.random() < 0.6:
            # Q4 holidays
            event_start = datetime(now.year, 10, 15)
            event_end = datetime(now.year, 12, 31)
            promotions.append({
                'sku': sku,
                'event_name': f"Q4 Holiday Campaign - {sku}",
                'event_type': 'holiday',
                'start_date': event_start.strftime('%Y-%m-%d'),
                'end_date': event_end.strftime('%Y-%m-%d'),
                'discount_percent': 15 + random.random() * 25,
                'expected_lift_percent': 25 + random.random() * 40,
                'status': 'planned'
            })

        # Black Friday/Cyber Monday
        if random.random() < 0.7:
            bf_date = datetime(now.year, 11, 25)  # Approximate Black Friday
            promotions.append({
                'sku': sku,
                'event_name': f"Black Friday Sale - {sku}",
                'event_type': 'flash_sale',
                'start_date': (bf_date - timedelta(days=2)).strftime('%Y-%m-%d'),
                'end_date': (bf_date + timedelta(days=2)).strftime('%Y-%m-%d'),
                'discount_percent': 30 + random.random() * 20,
                'expected_lift_percent': 50 + random.random() * 50,
                'status': 'planned'
            })

    return pd.DataFrame(promotions)

if __name__ == '__main__':
    print("Generating extended product catalog...")
    products_df = generate_products(120)
    products_df.to_csv('demo_products.csv', index=False)
    print(f"✓ Generated {len(products_df)} products -> demo_products.csv")

    print("Generating 36 months of sales history...")
    sales_df = generate_sales_history(products_df, months=36)
    sales_df.to_csv('demo_sales_history.csv', index=False)
    print(f"✓ Generated {len(sales_df)} sales records -> demo_sales_history.csv")

    print("Generating promotion events...")
    promo_df = generate_promotions(products_df)
    promo_df.to_csv('demo_promotions.csv', index=False)
    print(f"✓ Generated {len(promo_df)} promotion events -> demo_promotions.csv")

    print("\nDataset Summary:")
    print(f"  Products: {len(products_df)}")
    print(f"  Sales Records: {len(sales_df)} ({len(sales_df)//len(products_df)} months average)")
    print(f"  Promotion Events: {len(promo_df)}")
    print(f"\nFiles ready for import:")
    print(f"  - demo_products.csv")
    print(f"  - demo_sales_history.csv")
    print(f"  - demo_promotions.csv")
